submittal/src/main/java/com/kyriba/submittal/service/SpeakerServiceImpl.java

```diff
@@ -0,0 +1,145 @@
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.util.Pair;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.List;
+import java.util.stream.StreamSupport;
+
+import static java.util.stream.Collectors.toList;
+
+
+/**
+ * @author M-ABL
+ */
+@Service
+@Transactional
+@RequiredArgsConstructor
+public class SpeakerServiceImpl implements SpeakerService
+{
+  private final SpeakerRepository speakerRepository;
+  private final TopicRepository topicRepository;
+  private final RegistrationRepository registrationRepository;
+  private final KeyGenerator keyGenerator;
+
+
+  @Override
+  public Pair<Long, List<Long>> register(final String email,
+                                         final String password,
+                                         final PersonalInfo personalInfo,
+                                         final JobInfo jobInfo,
+                                         final List<TopicInfo> topics,
+                                         final long conferenceId)
+  {
+    if (speakerRepository.findByEmail(email).isPresent()) {
+      throw new DuplicateObjectException("Email is already in use");
+    }
+    Speaker speaker = new Speaker();
+    speaker.setStatus(SpeakerStatus.REGISTRATION);
+    speaker.setEmail(email);
+    speaker.setPassword(password);
+    speaker.setPersonalInfo(personalInfo);
+    speaker.setJobInfo(jobInfo);
+    speaker.setTopics(topics.stream()
+        .map(topicInfo -> new Topic(conferenceId, topicInfo, speaker))
+        .collect(toList()));
+
+    Speaker savedSpeaker = registrationRepository
+        .save(new Registration(speaker, keyGenerator.generate(KeyType.EMAIL)))
+        .getSpeaker();
+
+    return Pair.of(savedSpeaker.getId(), savedSpeaker.getTopics().stream().map(Topic::getId).collect(toList()));
+  }
+
+
+  @Override
+  public void finishRegistration(final String confirmationKey)
+  {
+    Registration registration = registrationRepository.findByConfirmationKey(confirmationKey)
+        .orElseThrow(() -> new ObjectNotFoundException("The given confirmation key does not exist"));
+    Speaker speaker = registration.getSpeaker();
+    speaker.activate();
+    registrationRepository.delete(registration);
+  }
+
+
+  @Override
+  @Transactional(readOnly = true)
+  public PersonalInfo getPersonalInfo(final long speakerId)
+  {
+    return findSpeakerById(speakerId).getPersonalInfo();
+  }
+
+
+  @Override
+  public void updatePersonalInfo(final long speakerId, final PersonalInfo personalInfo)
+  {
+    Speaker speaker = findSpeakerById(speakerId);
+    speaker.setPersonalInfo(personalInfo);
+  }
+
+
+  @Override
+  public List<Long> proposeTopics(final long speakerId, final List<TopicInfo> topics, final long conferenceId)
+  {
+    Speaker speaker = findSpeakerById(speakerId);
+    Iterable<Topic> createdTopics = topicRepository.saveAll(topics

```

Comment:
Why Iterable?

Reply:
why not? CRUDRepository returns Iterable.

Reply:
This is my bad, just simplify code using intermediate local variable for topics.